go/host/host.go

```diff
@@ -13,112 +14,101 @@ import (
+func NewHost(
+	config *config.HostConfig,
+	database *db.DB,
+	l1Client ethadapter.EthClient,
+	l1Publisher *l1.Publisher,
+	p2p P2PHostService,
+	enclaveClient common.Enclave,
+	logger gethlog.Logger,
+) hostcommon.Host {
+	// set up host metadata
 	hostIdentity := hostcommon.NewIdentity(config)
-	host := &host{
-		// config
-		config:  config,
-		shortID: common.ShortAddress(config.ID),
-
-		// services
-		services: hostServices,
-
-		// Initialize the host DB
-		db: database,
-
-		logger:         logger,
-		metricRegistry: regMetrics,
 
-		stopControl: stopcontrol.New(),
-	}
-	host.interrupter = breaker.Multiplex(
+	// set up control mechanisms
+	stopControl := stopcontrol.New()
+	interrupter := breaker.Multiplex(
 		breaker.BreakBySignal(
 			os.Kill,
 			os.Interrupt,
 		),
 	)
-	enclGuardian := enclave.NewGuardian(config, hostIdentity, hostServices, enclaveClient, database, host.interrupter, logger)
-	enclService := enclave.NewService(hostIdentity, hostServices, enclGuardian, logger)
-	l2Repo := l2.NewBatchRepository(config, hostServices, database, logger)
-	subsService := events.NewLogEventManager(hostServices, logger)
-
-	hostServices.RegisterService(hostcommon.P2PName, p2p)
-	hostServices.RegisterService(hostcommon.L1BlockRepositoryName, l1Repo)
-	hostServices.RegisterService(hostcommon.L1PublisherName, l1.NewL1Publisher(hostIdentity, ethWallet, ethClient, mgmtContractLib, l1Repo, logger))
-	hostServices.RegisterService(hostcommon.L2BatchRepositoryName, l2Repo)
-	hostServices.RegisterService(hostcommon.EnclaveServiceName, enclService)
-	hostServices.RegisterService(hostcommon.LogSubscriptionServiceName, subsService)
-
-	var prof *profiler.Profiler
-	if config.ProfilerEnabled {
-		prof = profiler.NewProfiler(profiler.DefaultHostPort, logger)
-		err := prof.Start()
-		if err != nil {
-			logger.Crit("unable to start the profiler: %s", log.ErrKey, err)
-		}
-	}
 
+	// create the Host controlled Services
+	l1RepoService := l1.NewL1Repository(l1Client, logger)
+	l2RepoService := l2.NewBatchRepository(config, p2p, database, logger)
+	logEventService := events.NewLogEventManager(enclaveClient, logger)
+	// todo review why there are 2 inits for the guardian ( guardian component + service )
+	// both exist at the same level and both have start/stop but the service should be the only available interface to the host
+	enclaveGuardianLib := enclave.NewGuardian(
+		config,
+		hostIdentity,
+		p2p,
+		l1Publisher,
+		l1RepoService,
+		l2RepoService,
+		logEventService,
+		enclaveClient,
+		database,
+		interrupter,
+		logger,
+	)
+	enclaveService := enclave.NewService(hostIdentity, p2p, enclaveGuardianLib, logger)
+
+	// log startup data for sanity control / tests
 	jsonConfig, _ := json.MarshalIndent(config, "", "  ")
 	logger.Info("Host service created with following config:", log.CfgKey, string(jsonConfig))
 
-	return host
+	return &host{
+		APIDBRepository:  database,
+		APIEnclaveClient: enclaveClient,
```

Comment:
> So currently this means the APIs are talking directly to the enclave. 
Yes, that avoids that host impl has to be a passthrough like the enclave impl.

> ut you're proposing I guess that eventually EnclaveService would implement APIEnclaveClient and it can delegate the methods to whichever enclave it wants to?

The proposal would be using this pattern and when we need the "host-hop" we just overload the method. Does that make sense ?